processing/src/main/java/io/druid/query/QueryToolChest.java

```diff
@@ -36,11 +46,81 @@
 public abstract class QueryToolChest<ResultType, QueryType extends Query<ResultType>>
 {
   /**
+   * @param filter Filter input to optimize
+   *
+   * @return Returns an optimized {@link DimFilter}
+   * Optimization is done by rewriting all ExtractionDimFilter of type {@link LookupExtractionFn} as a clause of literals {@link SelectorDimFilter}
+   */
+  public static DimFilter optimizeExtractionDimFilter(final DimFilter filter)
+  {
+    if (filter instanceof NotDimFilter) {
+      return Druids.newNotDimFilterBuilder().field(optimizeExtractionDimFilter(((NotDimFilter) filter).getField()))
```

Comment:
can the filters themselves have an optimize() method they call internally to avoid this gigantic if/else loop?